akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Actor.scala

```diff
@@ -183,6 +183,14 @@ object Actor {
 
     def onSignal(onSignal: PartialFunction[(ActorContext[T], Signal), Behavior[T]]): Behavior[T] =
       new BehaviorImpl.ImmutableBehavior(onMessage, onSignal)
+
+    def onResponse[M: ClassTag](handler: (ActorContext[T], M) ⇒ Behavior[T]): Immutable[T] = {
+      val x: (ActorContext[T], Any) ⇒ Behavior[T] = (ctx, msg) ⇒ msg match {
+        case m: M ⇒ handler(ctx, m)
+        case _    ⇒ onMessage(ctx, msg.asInstanceOf[T])
+      }
+      new Immutable[T](x.asInstanceOf[(ActorContext[T], T) ⇒ Behavior[T]])
+    }
```

Comment:
Is this really an improvement over `spawnAdapter`? We loose type safety and can still not tie in a per call context...

Reply:
See the description of the PR. As far as I can see, `spawnAdapter` cannot be used for temporary adapters without high risk of resource leak, and they require boilerplate of defining internal messages.

For per call context we would have the "safe" ask.

Reply:
I think I may have an idea to solve both now. I'll write down a note and try it out on wednesday.